File: project_euler_unified/problem_213/solution.py
import numpy as np

def solve():
    R, C = 30, 30
    N = R * C
    ROUNDS = 50

    def idx(r, c):
        return r * C + c

    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]

    deg = np.zeros(N, dtype=float)
    for r in range(R):
        for c in range(C):
            deg[idx(r, c)] = sum(
                1 for k in range(4) if 0 <= r + dr[k] < R and 0 <= c + dc[k] < C
            )

    T = np.zeros((N, N), dtype=float)
    for r in range(R):
        for c in range(C):
            i = idx(r, c)
            for k in range(4):
                nr, nc = r + dr[k], c + dc[k]
                if 0 <= nr < R and 0 <= nc < C:
                    T[i, idx(nr, nc)] = 1.0 / deg[i]

    Tn = np.linalg.matrix_power(T, ROUNDS)
    log_prod_not = np.sum(np.log1p(-Tn), axis=0)
    answer = np.sum(np.exp(log_prod_not))

    print(f"{answer:.6f}")

if __name__ == "__main__":
    solve()

File: project_euler_unified/problem_213/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    const int R = 30, C = 30;
    const int N = R * C;
    const int ROUNDS = 50;

    auto idx = [&](int r, int c) { return r * C + c; };

    int dr[] = {-1, 1, 0, 0};
    int dc[] = {0, 0, -1, 1};

    vector<int> deg(N);
    for (int r = 0; r < R; r++)
        for (int c = 0; c < C; c++) {
            int d = 0;
            for (int k = 0; k < 4; k++)
                if (r + dr[k] >= 0 && r + dr[k] < R && c + dc[k] >= 0 && c + dc[k] < C)
                    d++;
            deg[idx(r, c)] = d;
        }

    vector<double> prod_not(N, 1.0);

    for (int sr = 0; sr < R; sr++) {
        for (int sc = 0; sc < C; sc++) {
            vector<double> cur(N, 0.0), nxt(N, 0.0);
            cur[idx(sr, sc)] = 1.0;

            for (int t = 0; t < ROUNDS; t++) {
                fill(nxt.begin(), nxt.end(), 0.0);
                for (int r = 0; r < R; r++)
                    for (int c = 0; c < C; c++) {
                        int i = idx(r, c);
                        if (cur[i] == 0.0) continue;
                        double p = cur[i] / deg[i];
                        for (int k = 0; k < 4; k++) {
                            int nr = r + dr[k], nc = c + dc[k];
                            if (nr >= 0 && nr < R && nc >= 0 && nc < C)
                                nxt[idx(nr, nc)] += p;
                        }
                    }
                swap(cur, nxt);
            }

            for (int i = 0; i < N; i++)
                prod_not[i] *= (1.0 - cur[i]);
        }
    }

    double answer = 0.0;
    for (int i = 0; i < N; i++)
        answer += prod_not[i];

    cout << fixed << setprecision(6) << answer << endl;
    return 0;
}
